test/03_device-state.ts

```diff
@@ -510,6 +553,314 @@ export default () => {
 									api_heartbeat_state: DeviceOnlineStates.Offline,
 								});
 							});
+
+							describe('when the write cache expires while the device is on Timeout (eg API stops serving requests)', function () {
+								let device3: fakeDevice.Device;
+								const device3ChangeEventSpy = sinon.spy();
+
+								before(async function () {
+									device3 = await fakeDevice.provisionDevice(
+										admin,
+										applicationId,
+									);
+									stateMock.getInstance().on('change', (args) => {
+										if (device3.id === args.deviceId) {
+											device3ChangeEventSpy(args);
+										}
+									});
+									await fakeDevice.getState(
+										device3,
+										device3.uuid,
+										stateVersion,
+									);
+									await waitFor({
+										checkFn: () => device3ChangeEventSpy.called,
+									});
+									device3ChangeEventSpy.resetHistory();
+									await expectDeviceHeartbeat(device3.id, {
+										db: DeviceOnlineStates.Online,
+										cache: DeviceOnlineStates.Online,
+									});
+
+									await setTimeout(heartbeatTimeoutChangeInterval);
+									await waitFor({
+										checkFn: () => device3ChangeEventSpy.called,
+									});
```

Comment:
Do we need the `await setTimeout()` if we're executing `await waitFor()` directly after?

Reply:
What you say is correct. It seems that the mock device poll interval (thus how much time is needed to switch to Timeout) is small enough that `waitFor` is enough.
If we later increase the mock poll interval to a value that above the max default wait time that `waitFor` has, we then would need the setTimeout as well.
I will go with `waitFor` & leave a comment.